### ISSUE TITLE
filteredOn is not working with JUnitSoftAssertions

### ISSUE THREAD

**Issue by cajar1988** [joel-costigliola/assertj-core#512]
Here's sample test code:
```
//someList is type of java.util.List<T>

assertThat(someList)
  .filteredOn(v -> v.getValue().equals("someValue"))
  .hasSize(1);
```
This code works fine, but when I change it to:
```
@Rule
public final JUnitSoftAssertions softly = new JUnitSoftAssertions();

...

softly.assertThat(someList)
  .filteredOn(v -> v.getValue().equals("someValue"))
  .hasSize(1);
```

Then assert is done on not filtered list.

Also in debugger I've found message: Method threw 'java.lang.ClassCastException' exception. Cannot evaluate org.assertj.core.api.ListAssert$$EnhancerByCGLIB$$d9bf1c17.toString()

Furthermore, using soflty as `SoftAssertions softly = new SoftAssertions();` also works fine.
Problem occurs only with JUnit rule usage.

Version: 3.2.0

**Comment by joel-costigliola** [comment]
That's a bug, the reason being that `filter` operation create a new `ListAssert` with the filtered collection but the soft assertions work by proxying the initial `ListAssert` whcih remains unchanged.

Thanks for reporting this.

**Comment by joel-costigliola** [comment]
Fixed, will also be available for 3.3.0.
